### ISSUE TITLE
GENIE3 implementation

### ISSUE THREAD

**Issue by rojinsafavi** [aertslab/pySCENIC#362]
Hello!
I was wondering if any sort of normalization is applied to the final adj output (with the GENIE3 algorithm) across all the adjacencies?

To be more exact, I'd like to compute the adjacencies for 10 genes, and for each gene, I am using a different set of TFs. I have already modified the SCENIC code a bit to make it compatible with getting different inputs for each gene. My question is whether SCENIC does any global normalization or postprocessing on all the adjacencies that I need to be aware of? 

Thank you!

**Comment by SeppeDeWinter** [comment]
Hi @rojinsafavi 

No additional normalisation is applied after running GENIE3, the output is like the output you would get if you would run GENIE3 alone (https://github.com/aertslab/arboreto). 

Hope this helps.

Good luck.

Seppe

**Comment by rojinsafavi** [comment]
Thanks @SeppeDeWinter ,
I tried running GENIE3 directly from arboreto package, but I was getting some errors. I read the arboreto code where the GENIE3 is applied, and to me, it seemed like there is no normalization on the importance values after GENIE3 run is finished. I would very much appreciate it if I can get your confirmation on that!

basically, I want to know if the important values are gonna be any different in these to scenarios:



<img width="695" alt="Screen Shot 2022-01-26 at 9 56 03 AM" src="https://user-images.githubusercontent.com/19545198/151219423-3ace791a-3409-4e90-89df-5d590e593915.png">

Is my assumption right that the importance values for gene1 should be the same in both scenarios?

Thank you very much again @SeppeDeWinter

**Comment by SeppeDeWinter** [comment]
Hi @rojinsafavi 

Now I understand your question fully, thanks for the nice illustration.

What GENIE3 does under the hood is fit a RandomForrestRegressor (see https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestRegressor.html#sklearn.ensemble.RandomForestRegressor.fit for more info) with the TF matrix as input (predictor) and the target gene as output (i.e. will predict the expression vector of the gene based in the expression vector of all TFs). It repeats this process for each gene. 
After the model is fit, to get the importance values for each TF-gene link, it will get the feature importance for each gene, with the features being the TFs (feature_importances_ attribute in scikit learn). 

The core code in arboreto is this: 

```python
regressor.fit(tf_matrix, target_gene_expression)
```

see: https://github.com/aertslab/arboreto/blob/2f475dca08f47a60acc2beb8dd897e77b7495ca4/arboreto/core.py#L138

i.e. in your example, top: it will try to predict the expression of gene1 based on the expression of TF1, TF2 and TF3, next it will try to predict the expression of gene 2 based on the expression of TF1, TF2 and TF3, ... The importance value of TF1 to gene1 is the feature importance of TF1 for the prediction of gene1, The importance value of TF2 to gene1 is the feature importance of TF2 for the prediction of gene1, ...

I hope from this it is clear that if you only use a target gene matrix with a single gene (compared to a matrix with more genes) that the importance values for that single gene wont change*.

So to answer your question

**Is my assumption right that the importance values for gene1 should be the same in both scenarios?**
Yes.

Two notes:

1. (*) it can be that the values wont be exactly the same because there is some randomness in the algorithm, being sure that the same seed it used in both cases is thus key,
2. If you would do the reverse, i.e. change the TF matrix instead of the gene matrix, the values would in fact change because you've changed your predictor values (importance scores are relative to which features are included for the prediciton).

I hope this is clarifying. 

Best,

Seppe

**Comment by rojinsafavi** [comment]
Thank you so much!
